### PR TITLE
First attempt at networking (pretty ugly)

### PR DESCRIPTION
**Pull request by ZaneMODell**
Due to various issues we had with developer operating systems and libraries, we were not able to make something that is very cohesive or well put together, but something that works at a very low level.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -4,6 +4,7 @@ build
 # clion cmake build
 cmake-build-debug
 cmake-build-release
+CMakeLists.txt
```
Comment (amantham20) [inline .gitignore:7]: ya might wanna change that

File: source/Interfaces/NetWorth/NetworkInterface.hpp [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+#include <SFML/Network/Packet.hpp>
+
+#include "../../core/InterfaceBase.hpp"
+
+using namespace sf;
+
+namespace cse491 {
+    namespace netWorth{
+    /**
+     * TODO: Delete or incorporate
```
Comment (mercere99) [inline source/Interfaces/NetWorth/NetworkInterface.hpp:22]: It looks like this is used in the Client and Server interfaces, so I assume this is an old comment?

File: source/Interfaces/NetWorth/NetworkInterface.hpp [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+     */
+        class NetworkingInterface{
+        private:
+
+        protected:
+            UdpSocket mSocket; ///The socket we are going to make our connection
```
Comment (mercere99) [inline source/Interfaces/NetWorth/NetworkInterface.hpp:28]: With the new format we just discussed in class, I think member variables can still begin with m, but we should probably change these to something like `m_socket` for consistency.

File: source/Interfaces/NetWorth/NetworkInterface.hpp [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+             */
+            virtual void SendPacket(Packet packet, IpAddress destAddr, const unsigned short port){
+                mSocket.send(packet, destAddr, port);
+            }
+            /**
+             * Starts the connection by receiving the first packet
+             * @return the IP of the server
+             */
+            virtual IpAddress ReceivePacket(){
+                char buffer[1024];
```
Comment (mercere99) [inline source/Interfaces/NetWorth/NetworkInterface.hpp:66]: I'd recommend changing this over to a standard library array.  Specifically, it would look like:

```
std::array<char, 1024> buffer;
```

It would otherwise behave the same, but would have some member function to help out, like `buffer.size()`.

File: source/Interfaces/NetWorth/client/CMakeLists.txt
Comment (mercere99) [inline source/Interfaces/NetWorth/client/CMakeLists.txt:1]: I'm going to leave it to @FergusonAJ to comment on the CMake aspects of this pull request, but I don't see any obvious issues here.

File: source/Interfaces/NetWorth/client/ClientInterface.hpp [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+using namespace sf;
+
+namespace cse491 {
+    namespace netWorth{
+    /**
+     * The interface of our client that will be interacting and connection with our server
+     */
+        class ClientInterface {
+        private:
+            std::unique_ptr<TrashInterface> mTrash;     /// Display interface
```
Comment (mercere99) [inline source/Interfaces/NetWorth/client/ClientInterface.hpp:28]: You want to minimize the amount of information that you send across the network to just the bare minimum that an interface might need.  As such, you probably don't want to use a TrashInterface for the client side, but something that you build on your own (perhaps initially based on the TrashInterface) that requires only the more limited data that you decide to send it.  I think in the long term, this will be easier for you where you process down input data on the server side and send it along to the client.

All of that said, I don't see where you're actually setting up the client interface...

File: source/Interfaces/NetWorth/client/ClientInterface.hpp [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+                        std::cout << "Failure to receive" << std::endl;
+                        return;
+                    }
+                    recv_pkt >> recv_str;
+                    std::cout << std::endl << recv_str;
+
+                    valid_input = false;
+                    while (!valid_input) {
+                        do {
+                            std::cin >> input;
```
Comment (mercere99) [inline source/Interfaces/NetWorth/client/ClientInterface.hpp:126]: I don't think you need a loop around this `std::cin`; while we certainly want to move to a real-time interface, `cin` should wait for user input before continuing.

File: source/Interfaces/NetWorth/client/ClientInterface.hpp [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+                recv_pkt >> str;
+                std::cout << str << std::endl;
+
+                return true;
+            }
+
+            /**
+             * Game loop
+             */
+            void Loop() {
```
Comment (mercere99) [inline source/Interfaces/NetWorth/client/ClientInterface.hpp:85]: I'd push for a more descriptive name for this function than just `Loop()` since that's as generic as it comes.  At least call it `GameLoop()` which would provide some more information.

This function is also quite long; I'd recommend building a set of helper functions to make it more readable.  For example, you can have `GameLoop_Setup()`, `GameLoop_GetInput()`, `GameLoop_ProcessInput()`, `GameLoop_SendAction()`, etc.  I'm happy to provide more detailed suggestions if that would be helpful.

File: source/Interfaces/NetWorth/server/ServerInterface.hpp [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+             * @param item_set the items that may be apart of the grid
+             * @param agent_set the agents that may be apart of the grid
+             * @return the grid that will be sent to the client
+             */
+            static Packet GridToPacket(const cse491::WorldGrid & grid, const cse491::type_options_t & type_options,
+                                const cse491::item_set_t & item_set, const cse491::agent_set_t & agent_set)
+            {
+                std::vector<std::string> packet_grid(grid.GetHeight());
+
+                // Load the world into the symbol_grid;
```
Comment (mercere99) [inline source/Interfaces/NetWorth/server/ServerInterface.hpp:72]: Note that `WorldGrid` already has member functions `Serialize(std::ostream & os)` and `Deserialize(std::istream & is)`.  I think these might do the trick for you, but we can also alter them to simply work with strings rather than reading and writing to streams.

Comment (mercere99) [reply source/Interfaces/NetWorth/server/ServerInterface.hpp:72]: Ahhh..  you're not trying to serialize the whole grid, just generate the output for the user.  That makes more sense and should work well.

File: source/Interfaces/NetWorth/server/networkingworld.hpp
Comment (mercere99) [inline source/Interfaces/NetWorth/server/networkingworld.hpp:1]: Ideally a network interface will work with any world.  Is this just a test for you, or are you going to be limited to specific world types.  I think this can be done by having the server-side be an Interface and the client side takes all of the information from it and returns user inputs.  All-in-all, the world shouldn't need to know what kind of interface it is.

That said, if there is specific info that you need a world to manage, we may want to consider modifying the `WorldBase` API to ensure that all worlds will have whatever is needed.

File: source/Interfaces/NetWorth/server/servermain
Comment (mercere99) [inline source/Interfaces/NetWorth/server/servermain:1]: Please don't put binary files in the pull requests.

File: source/Interfaces/NetWorth/server/servermain.cpp [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+    std::shared_ptr<cse491::netWorth::ServerInterface> serverInterface = std::make_shared<cse491::netWorth::ServerInterface>();
+
+    world->SetServer(serverInterface);
+
+    UdpSocket * serverSocket = serverInterface->GetSocket();
+
+    std::cout << "Server IP Address: " << sf::IpAddress::getLocalAddress() << std::endl;
+
+    //Establish an initial connection
```
Comment (mercere99) [inline source/Interfaces/NetWorth/server/servermain.cpp:36]: Can establishing the connection for in the network interface constructor?  Then it should be possible to set it up with multiple interfaces without them stepping on each other's toes (I think...)  It will also keep everything more modular.

File: source/Interfaces/NetWorth/testmain
Comment (mercere99) [inline source/Interfaces/NetWorth/testmain:1]: Remove from pull request.

File: source/core/WorldGrid.hpp
Comment (mercere99) [inline source/core/WorldGrid.hpp:1]: I think the changes here are just for debugging?  They should not be part of the pull request.

File: source/simple
Comment (mercere99) [inline source/simple:1]: Remove from PR.

File: tests/build/.gitignore
Comment (mercere99) [inline tests/build/.gitignore:1]: I'm not sure about this change; it looks like it changes from ignoring all files in  `tests/build/` to only ignoring files that begin with a dash?

**Review by mercere99 — CHANGES_REQUESTED** [review]
This is a solid start on the networking interface.  I added some comments and suggestions throughout.

I think the main thing to lock in on is the style of the interface and how to simultaneously make it flexible and modular.  In this case, I think the interface should be treated like any other interface on the server side, but internal to the interface it should process the data to send to the client and collect the results.  I don't think that main or world should need to do anything special to make this work; most of the networking should be able to be setup in the constructor or related helper functions.

The more challenging piece will be to make sure multiple network interfaces can run simultaneously, but let's get one going smoothly before putting too much effort into that direction.

The main changes for now are the basic cleanup on the interface so that it can work with existing worlds that other teams are making, and removing some files that were accidentally added to pull request.

Good job with naming, style, and building on a modular API.  Initial tests look good!

File: source/Interfaces/NetWorth/NetworkInterface.hpp (hunk 2 above)
Comment (ZaneMODell) [reply source/Interfaces/NetWorth/NetworkInterface.hpp:22]: Currently this class is not being used, as it is meant to be a parent class for both the client and server interface, however the class itself is not being used as of this PR. It will be incorporated in upcoming versions of the module but we wanted to leave a TODO there so we did not forget that.

File: .gitignore
[the remark this answers is not in the document]
Comment (ZaneMODell) [reply .gitignore:1]: Can you clarify a bit on this? Do you want us to remove the CMakeLists.txt from the .gitignore? Are there any other specific changes you are looking for?

File: source/Interfaces/NetWorth/NetworkInterface.hpp (hunk 4 above)
Comment (ZaneMODell) [reply source/Interfaces/NetWorth/NetworkInterface.hpp:66]: The code used here was given from the SFML documentation and changing the type to std::array<char, 1024> does not match the type of a parameter for a socket receive function that the SFML API uses. Not sure what to do to change this other than doing some sort of potential strange type-casting.

File: source/Interfaces/NetWorth/client/ClientInterface.hpp (hunk 5 above)
Comment (ZaneMODell) [reply source/Interfaces/NetWorth/client/ClientInterface.hpp:28]: I believe we should have actually removed this from our current PR, as it is not being used. The intention was that the client would have that assigned at the beginning so that it when it initially connects to the server it would be given the interface, but due to a lack of serialization at the moment we were stuck on how to perform this.

File: source/Interfaces/NetWorth/server/networkingworld.hpp
Comment (ZaneMODell) [reply source/Interfaces/NetWorth/server/networkingworld.hpp:1]: Yeah, the goal is to have it work with any, or at the very least multiple kinds of worlds. This was just a test so we wouldn't change any other world files or manipulate anything that had already been written.

File: source/Interfaces/NetWorth/NetworkInterface.hpp (hunk 4 above)
Comment (mercere99) [reply source/Interfaces/NetWorth/NetworkInterface.hpp:68]: If you use std::array, you can always call `.data()` on it to convert it back to a regular array.  Either way should be fine though.

Comment (ZaneMODell) [reply source/Interfaces/NetWorth/NetworkInterface.hpp:68]: Ok, yeah that seems to work for now, we aren't actually using that class at the moment but we will make the change so that when our class hierarchy is complete we have that additional functionality.

File: tests/build/.gitignore
Comment (EthanRylko) [reply tests/build/.gitignore:1]: Honestly not sure how I ended up doing this, seems like I must've accidently brushed a key when I was working on tests.

File: source/Interfaces/NetWorth/client/ClientInterface.hpp (hunk 7 above)
Comment (EthanRylko) [reply source/Interfaces/NetWorth/client/ClientInterface.hpp]: Ended up making a GameLoop_GetInput() function, that was really helpful.

Also added some helper functions in the NetworkingInterface class. Every time I would send or receive a packet, a status code is returned and I would check it and print an error if something was wrong. I moved the check and error message into one function and made it return a bool to check for success (I still have to quit the loop if there is an error). Having the if statements for each send/receive looked pretty bulky. Can probably do more helper functions later.
